servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/PayloadSizeLimitingHttpRequesterFilter.java

```diff
@@ -0,0 +1,109 @@
+import io.servicetalk.concurrent.api.Single;
+import io.servicetalk.http.api.FilterableStreamingHttpClient;
+import io.servicetalk.http.api.FilterableStreamingHttpConnection;
+import io.servicetalk.http.api.HttpExecutionStrategy;
+import io.servicetalk.http.api.HttpExecutionStrategyInfluencer;
+import io.servicetalk.http.api.StreamingHttpClientFilter;
+import io.servicetalk.http.api.StreamingHttpClientFilterFactory;
+import io.servicetalk.http.api.StreamingHttpConnectionFilter;
+import io.servicetalk.http.api.StreamingHttpConnectionFilterFactory;
+import io.servicetalk.http.api.StreamingHttpRequest;
+import io.servicetalk.http.api.StreamingHttpRequester;
+import io.servicetalk.http.api.StreamingHttpResponse;
+
+import java.util.function.Function;
+
+import static io.servicetalk.http.api.HttpExecutionStrategies.offloadNone;
+
+/**
+ * Limits the response payload size for a request. The filter will throw an exception which may result in
+ * stream/connection closure.
+ */
+public final class PayloadSizeLimitingHttpRequesterFilter implements
+                        StreamingHttpClientFilterFactory, StreamingHttpConnectionFilterFactory,
+                        HttpExecutionStrategyInfluencer {
+    private final int maxResponsePayloadSize;
+
+    /**
+     * Create a new instance.
+     * @param maxResponsePayloadSize The maximum response payload size allowed.
+     */
+    public PayloadSizeLimitingHttpRequesterFilter(int maxResponsePayloadSize) {
+        if (maxResponsePayloadSize < 0) {
+            throw new IllegalArgumentException("maxResponsePayloadSize: " + maxResponsePayloadSize + " (expected >0)");
+        }
+        this.maxResponsePayloadSize = maxResponsePayloadSize;
+    }
+
+    @Override
+    public StreamingHttpClientFilter create(final FilterableStreamingHttpClient client) {
+        return new StreamingHttpClientFilter(client) {
+            @Override
+            protected Single<StreamingHttpResponse> request(final StreamingHttpRequester delegate,
+                                                            final StreamingHttpRequest request) {
+                return applyLimit(request, delegate::request);
+            }
+        };
+    }
+
+    @Override
+    public StreamingHttpConnectionFilter create(final FilterableStreamingHttpConnection connection) {
+        return new StreamingHttpConnectionFilter(connection) {
+            @Override
+            public Single<StreamingHttpResponse> request(final StreamingHttpRequest request) {
+                return applyLimit(request, super::request);
+            }
+        };
+    }
+
+    @Override
+    public HttpExecutionStrategy requiredOffloads() {
+        return offloadNone();
+    }
+
+    private Single<StreamingHttpResponse> applyLimit(
+            StreamingHttpRequest request, Function<StreamingHttpRequest, Single<StreamingHttpResponse>> delegator) {
+        return delegator.apply(request).map(response -> response.transformPayloadBody(publisher ->
```

Comment:
Have you considered using `transformMessageBody`? While it's more advanced, it will result in less overhead and won't modify `PayloadInfo` flags. For a filter that we provide, it's justified to use more advanced transformation method.

Reply:
Sure I can do this.